### ISSUE TITLE
How to store a part of the output of a Flow in a Response in a data store

### ISSUE THREAD

**Issue by JaGeo** [materialsproject/jobflow#393]
I am currently struggling to find a way to store a part of the output of Flow, that is  created with a Response object, in the data store. Is this possible in general? 
If I can pass it via "stored_data": could we then document it in a better way?

**Comment by JaGeo** [comment]
I am not sure I can easily create a minimal example for it. I would like to store a part of the output in this job here ("forces", "displaced_structures", etc.) in the data store:

https://github.com/JaGeo/atomate2/blob/af4af8e0363365c2ce774d5251386b99d94aac16/src/atomate2/common/jobs/phonons.py#L315

I am trying to run larger cells with many displacements which leads to a overflow.

**Comment by JaGeo** [comment]
A minimal could look like this:
`from jobflow import job, Flow, Response, run_locally

@job
def add(a, b):
    return a + b

@job
def large_add(a,b):
    job0 = add(a, b)
    job1 = add(job0.output, 1)
    output = {}
    output["result"] = job1.output

    flow = Flow([job0, job1], output)
    return Response(flow)


run_locally(Flow(large_add(0,1)))`

I suspected I could use `stored_data={"data": ["result"]}` in Response.

**Comment by utf** [comment]
Ah, I see. The `stored_data` is not actually to do with the data store. It is a copy from the FireWorks stored_data option, which stores extra information alongside the launch information in the database.

To achieve this, you need to set the data option on the job. E.g., 

```python

@job(data=True)
def add(a, b):
    return a + b

def large_add(a,b):
    job0 = add(a, b)
    job1 = add(job0.output, 1)
    output = {}
    output["result"] = job1.output

    flow = Flow([job0, job1], output)
    return Response(flow)
```

The output of the Flow itself doesn't get stored in the database, only references to the previous outputs get stored (literally, we store the output reference in the database).

**Comment by JaGeo** [comment]
Thanks! However, there could be still a lot of references that I might want to store in a datastore:
Follow up question:
How could I store something in the datastore that is not part of the output of add? e.g., 

```
@job(data=True)
def add(a, b):
    return a + b

def large_add(a,b):
    job0 = add(a, b)
    job1 = add(job0.output, 1)
    output = {}
    output["result"] = job1.output
    output["input0"] = a

    flow = Flow([job0, job1], output)
    return Response(flow)
```

**Comment by utf** [comment]
I'm actually a little confused by your code. You're returning a response object from a function which is not a job. Should `large_add` have the `@job` decorator?

**Comment by JaGeo** [comment]
Ah, yes! It should have! Sorry!

**Comment by utf** [comment]
Ok! The other confusing thing is that you are returning a Response object with a Flow set to the output. E.g., your code is equivalent to `Response(output=flow)`. Should this instead be a dynamic flow, e.g., returning a replace: `Response(replace=flow)`. This second option probably does what you want. E.g., you can access the output of large_add as follows:

```
my_job = large_add(1, 2)
my_job.output["input0"]
```

If that is the case, you should be able to use the standard way of storing objects in the data store. E.g., 

```python
@job(data=["input0"])
def large_add(a,b):
    job0 = add(a, b)
    job1 = add(job0.output, 1)
    output = {}
    output["result"] = job1.output
    output["input0"] = a

    flow = Flow([job0, job1], output)
    return Response(replace=flow)
```

In the above code, when you perform the replace, three jobs get added. The first two are the jobs in your flow (e.g., `job0` and `job1`). The third job is a [`store_inputs` job](https://github.com/materialsproject/jobflow/blob/c9d9e811a531a35d68a147863076550e856c3e41/src/jobflow/core/job.py#L1270) to add the output of your flow to the JobStore. Note, the only time *flow* outputs get stored is when then are part of a dynamic workflow, normally only job outputs get stored.

As an aside, the `store_inputs` naming is a bit confusing. It is storing the *outputs* of your flow but these are in fact the *inputs* to the `store_inputs` function, hence the name. Perhaps this could be changed.

The `store_inputs` job *should* inherit the same data store settings as the job which triggered the dynamic behaviour. E.g., in this case it would be the `large_add` job. However, looking at the code, the `store_inputs` job currently only inherits the metadata and output schema but not the data store settings. E.g., see here: https://github.com/materialsproject/jobflow/blob/c9d9e811a531a35d68a147863076550e856c3e41/src/jobflow/core/job.py#L1319-L1324

This is definitely a bug and should be fixed. I think the fix is to also copy the `job._kwargs` key across. These kwargs detail which objects get stored in which stores. E.g., the fix is to add a line here: https://github.com/materialsproject/jobflow/blob/c9d9e811a531a35d68a147863076550e856c3e41/src/jobflow/core/job.py#L1323
```python
store_output_job._kwargs = current_job._kwargs
```

Would you be willing to see if making that change fixes your issue?

**Comment by JaGeo** [comment]
Thanks! Yes, of course, this flow should have replaced this job! I made an error in the minimal example. Sorry for this. I should have been more careful.

I can test this after my vacation! I think this is something we will need to be fixed in future implementations. 

I tried adding the configuration for the datastore to the `large_add` job as well and also didn't see any change. Thanks for confirming that this is a bug.
